### DISCUSSION TITLE
How to change span name for handle request spans?

### DISCUSSION THREAD

**Discussion by andrew-demb** [vercel/next.js#59645] [Help] [answered]
### Summary

OpenTelemetry docs states, that span name should not contain variables to avoid high cardinality [1]. But NextJS uses a default span name like "GET /requested/pathname".

I read the documentation [2] which states, that NextJS can use route patterns as a part of span tags and mentioned the same wording ("next route") for root span about handling requests.
> next.route - The route pattern of the request (e.g., /[param]/user).
But searching in code about generating span logic leads to this - [3], which uses `req.url` (not route) and I can't find a way how to override span name.

**Questions**:
1. How can I change the default logic?
2. Is it intended - to use the request URL in the span name, but not the route name/pattern? 

[1] https://github.com/open-telemetry/opentelemetry-specification/blob/v1.28.0/specification/trace/api.md#span
![image](https://github.com/vercel/next.js/assets/12499813/1ef28d69-7488-44d9-a41e-9cf0102aeb9c)

[2] https://github.com/vercel/next.js/blob/v14.0.5-canary.13/docs/02-app/01-building-your-application/06-optimizing/08-open-telemetry.mdx#L218
[3] https://github.com/vercel/next.js/blob/05eb8114f9595561f9271dace5f6da55feac391d/packages/next/src/server/base-server.ts#L791-L797

**Comment by devjiwonchoi** [comment, marked as answer]
> HTTP span names SHOULD be {method} {target} if there is a (low-cardinality) target available. If there is no (low-cardinality) {target} available, HTTP span names SHOULD be {method}.

x-ref: [OTel docs](https://opentelemetry.io/docs/specs/semconv/http/http-spans/#:~:text=HTTP%20span%20names%20SHOULD%20be%20%7Bmethod%7D%20%7Btarget%7D%20if%20there%20is%20a%20(low%2Dcardinality)%20target%20available.%20If%20there%20is%20no%20(low%2Dcardinality)%20%7Btarget%7D%20available%2C%20HTTP%20span%20names%20SHOULD%20be%20%7Bmethod%7D.)

As the OTel docs recommend, changed the behavior at https://github.com/vercel/next.js/pull/75416 to use only the {method} if the available {target} is a high cardinality URL. Sorry for the late response, and thank you for reporting the issue!
